### PR TITLE
Functional tests

### PR DESCRIPTION
**Pull request by rachidlamouri**
Adding functional tests for these features
- random words
- turn counters
- reset buttons

Additional project tasks were added for these features
- card colors
- admin mode

### CODE REVIEW

File: tests/functional/support/hooks.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  if (result.status === Status.FAILED) {
+    const filepath = `${REPORTS_DIR}${pickle.name}.png`.replace(/ /g, '_');
+    await this.page.screenshot({ path: filepath });
+    console.log(`Error screenshot: ./${filepath}`); // eslint-disable-line no-console
+  }
+});
+
+After(async function () {
+  await this.page.close();
+  return new Promise((resolve) => setTimeout(resolve, 2000));
```
Comment (mollyhenderson) [inline tests/functional/support/hooks.js:47]: :wave: why is this 2sec delay here? (Maybe we need a comment?)

File: tests/functional/support/hooks.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
+global.nativePromise = Promise;
+const fs = require('fs');
+global.Promise = require('bluebird');
```
Comment (mollyhenderson) [inline tests/functional/support/hooks.js:3]: Don't override the global Promise plz

File: tests/functional/support/steps.js [hunk 3]
```diff
@@ -0,0 +1,183 @@
...
+defineStep('I see 25 unique random words', async function () {
+  const cards = await this.model.cards();
+  const words = await Promise.mapSeries(cards, (card) => card.word());
+  expect(words).to.have.lengthOf(25);
+
+  words.forEach((word) => {
+    expect(word, `"${word}"`).to.be.jsonSchema({
+      type: 'string',
+      minLength: 1,
+      pattern: '^[A-Z][A-Z ]*[A-Z]$',
```
Comment (mollyhenderson) [inline tests/functional/support/steps.js:49]: Why is there a space? Also, this regex implies a min length of 2

File: tests/functional/support/steps.js [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+  }
+  usedWords.push(...words);
+
+  await Promise.each(
+    cards,
+    async (card) => {
+      const topHalf = await card.topHalf();
+      const bottomHalf = await card.bottomHalf();
+
+      expect(await topHalf.word()).to.equal(await bottomHalf.word());
```
Comment (mollyhenderson) [inline tests/functional/support/steps.js:65]: Should probably have some info in the error message.

File: tests/functional/support/steps.js [hunk 5]
```diff
@@ -0,0 +1,183 @@
...
+    case 'top':
+      resetButton = await this.model.topResetButton();
+      break;
+    case 'bottom':
+      resetButton = await this.model.bottomResetButton();
+      break;
+    default: throw Error(`unexpected "${label}"`);
+  }
+
+  await resetButton.element.click();
```
Comment (mollyhenderson) [inline tests/functional/support/steps.js:37]: We shouldn't be accessing `element` here.

File: tests/functional/features/turnCounters.feature [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    Examples:
+      | playerId | turnCount | turnType   | remainingCount | indicesClause                 |
+      | 0        | 1         | successful | 8              |                               |
+      | 0        | 3         | bystander  | 6              | on cards 13, 14, 16           |
+      | 0        | 5         | failed     | 4              | on cards 1, 6, 9, 17, 23      |
+      | 1        | 2         | successful | 7              |                               |
+      | 1        | 4         | bystander  | 5              | on cards 8, 13, 17, 23        |
+      | 1        | 6         | failed     | 3              | on cards 3, 5, 10, 12, 14, 18 |
+
+  Scenario Outline: All turns taken
```
Comment (mollyhenderson) [inline tests/functional/features/turnCounters.feature:23]: Looks like this *might* not be running. Should this just be a `Scenario`?

File: tests/functional/features/turnCounters.feature [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+
+  Scenario Outline: Removing turn counters
+    Given there are 9 side counters
+    And player <playerId> takes 3 <turnType> turns <indicesClause>
+    When player <playerId> removes 1 <turnType> turn counter <indexClause>
+    Then player <playerId> has 2 <turnType> turn counters
+    And there are 7 side counters
+
+    Examples:
+      | playerId | turnType   | remainingCount | indicesClause      | indexClause |
```
Comment (mollyhenderson) [inline tests/functional/features/turnCounters.feature:40]: `remainingCount` is unused

File: tests/functional/support/steps.js [hunk 8]
```diff
@@ -0,0 +1,183 @@
...
+    _.range(turnCount),
+    async () => {
+      await addCounterButton.element.click();
+    },
+  );
+});
+
+defineStep(/^player (\d+) takes (\d+) (bystander|failed) turns on cards (.*)$/, async function (playerId, turnCount, turnType, indicesClause) {
+  const cards = await this.model.cards();
+  const cardIndicesText = indicesClause.replace(/on cards /, '');
```
Comment (mollyhenderson) [inline tests/functional/support/steps.js:105]: Looks like this line might not be doing anything (`indicesClause` shouldn't contain `on cards` at this point)

**Review by mollyhenderson — CHANGES_REQUESTED** [review]
Request the changes! 🦑

**Comment by rachidlamouri** [comment]
This is a year old, and this project is too far into my backlog for this to be updated any time soon. I'll reopen if I ever get back to it.

**Pull request closed without merging**